src/data/fintype/card_inj.lean

```diff
@@ -0,0 +1,165 @@
+        ext, rw function.funext_iff at eq,
+        specialize eq 0, rwa [fin.cons_zero, fin.cons_zero] at eq },
+      -- simp is getting hung up on `bex_def` here sadly, so have to do it manually
+      have mem_extended : ∀ {g : fin n.succ ↪ β}, g ∈ extended → ∃ a ∈ poss_vals, extend ⇑f a = g,
+        intros g g_extended, simp only [extended, mem_map] at g_extended,
+        obtain ⟨⟨a, a_poss⟩, -, g_extended⟩ := g_extended,
+        simp only [function.embedding.coe_fn_mk, subtype.coe_mk] at g_extended,
+        refine ⟨a, a_poss, _⟩, rw ←g_extended, simp,
+
+      have : |extended| = ‖β‖ - n, by simp [extended, poss_vals, card_sdiff, card_univ],
+
+      rw ←this, apply card_le_of_subset, rintros g g_extended,
+      simp only [extend, mem_equiv],
+      use g 0, ext t, revert t,
+      refine fin.induction (by rw fin.cons_zero) _,
+      rintros i -,
+      obtain ⟨k, untouched, g_extended⟩ := mem_extended g_extended,
+      rw ←g_extended, simp [extend] } },
+
+  rw [←card_univ, all_injf_covered, card_bUnion], swap, -- card_bUnion has a disjointness req
+  { rintros g - j - g_ne_j, rw disjoint_iff_ne, intros a a_equiv b b_equiv,
+    intro a_eq_b, apply g_ne_j, rw mem_equiv at a_equiv b_equiv,
+    obtain ⟨k₁, a_equiv⟩ := a_equiv, obtain ⟨k₂, b_equiv⟩ := b_equiv,
+    simp only [extend] at a_equiv b_equiv, subst a_eq_b, rw ←b_equiv at a_equiv,
+    apply_fun fin.tail at a_equiv, repeat { rw fin.tail_cons at a_equiv },
+    ext, rw a_equiv },
+  unfold nat.desc_fac,
+
+  suffices : ‖fin n ↪ β‖ * (‖β‖ - n) = (‖β‖ - n.succ + n + 1) * nat.desc_fac (‖β‖ - n.succ) n,
+  { simpa [equiv_class_size, card_univ] },
+
+  rw hn (nat.lt_of_succ_le h).le,
+  set t := ‖β‖ - n.succ with ht,
+  have : ‖β‖ - n = t.succ,
+  { rw [ht, nat.succ_eq_add_one, ←nat.sub_sub_assoc, nat.succ_sub_one],
+    exact h, exact nat.succ_pos _ },
+  rw [this, mul_comm, nat.succ_desc_fac]
+end
+
+/- Establishes the cardinality of the type of all injections, if any exist.  -/
+@[simp] theorem card_inj {α β} [fintype α] [fintype β] [decidable_eq α] (h : ‖α‖ ≤ ‖β‖)
+  : ‖α ↪ β‖ = (nat.desc_fac (‖β‖ - ‖α‖) ‖α‖) :=
+begin
+  trunc_cases fintype.equiv_fin α with eq,
+  rw fintype.card_congr (function.embedding.equiv eq (equiv.refl β)),
+  exact card_inj_aux _ _ h,
+end
+
+/-- If `‖β‖ < ‖α‖` there is no embeddings `α ↪ β`. This is the pigeonhole principle. -/
+@[simp] theorem card_inj' {α β} [fintype α] [fintype β] (h : ‖β‖ < ‖α‖) : ‖α ↪ β‖ = 0 :=
+begin
+  rw card_eq_zero_iff, intro f,
+  obtain ⟨x, y, eq, fne⟩ := fintype.exists_ne_map_eq_of_card_lt f h,
+  have := f.injective fne, contradiction
+end
+
+theorem card_inj'' {α β} [fintype α] [fintype β] :
+  ‖α ↪ β‖ = if ‖α‖ ≤ ‖β‖ then nat.desc_fac (‖β‖ - ‖α‖) ‖α‖ else 0 :=
+begin
+  split_ifs with h,
+    exact card_inj h,
+    exact card_inj' (not_le.mp h)
+end
+
+end fintype
+
+-- just realised; is it worth registering `subsingleton` instances for `‖α ↪ β‖`
+-- for when they either have equal cards or `α` is empty?
```

Comment:
> for when they either have equal cards

No, because that's not a true statement

> or `α` is empty?

We don't have a typeclass for "empty" yet.

Reply:
my brain clearly melted when I wrote that comment, thank you